File: Kingdom-Fall-UI-Enabled/src/main/java/application/Sauvegarde.java
package application;

import controller.FenetreAppController;
import controller.OptionsController;
import javafx.stage.FileChooser;

import java.io.*;

public class Sauvegarde
{
    static final String SAVE_DIRECTORY_PATH = System.getProperty("user.home") + "/games/Kingdom-Fall/sauvegardes";

    public static void sauvegarderJeu(Jeu jeu, String nomJoueur)
    {
        FileChooser fileChooser = new FileChooser();
        fileChooser.setTitle("Charger jeu");
        fileChooser.getExtensionFilters().add(new FileChooser.ExtensionFilter("Jeux", "*.jeu"));
        File directory = new File(SAVE_DIRECTORY_PATH);
        System.out.println(directory.getAbsolutePath());
        if(!directory.exists())
            directory.mkdirs();
        fileChooser.setInitialDirectory(directory);
        fileChooser.setInitialFileName(nomJoueur + ".jeu");
        File file = fileChooser.showSaveDialog(null);

        try (ObjectOutputStream out = new ObjectOutputStream(new FileOutputStream(file))) {
            out.writeObject(jeu);
            System.out.println("Jeu sauvegardé avec succès.");
        } catch (IOException e) {
            e.printStackTrace();
            System.out.println("Erreur lors de la sauvegarde.");
        }
    }

    public static Jeu chargerJeu()
    {
        FileChooser fileChooser = new FileChooser();
        fileChooser.setTitle("Charger jeu");
        fileChooser.getExtensionFilters().add(new FileChooser.ExtensionFilter("Jeux", "*.jeu"));
        File directory = new File(SAVE_DIRECTORY_PATH);
        System.out.println(directory.getAbsolutePath());
        if(!directory.exists())
            directory.mkdirs();
        fileChooser.setInitialDirectory(directory);
        File file = fileChooser.showOpenDialog(FenetreAppController.getSingleton().stageApp);

        try (ObjectInputStream in = new ObjectInputStream(new FileInputStream(file))) {
            Jeu jeu = (Jeu) in.readObject();
            OptionsController.renommerJoueur(jeu.getJoueur().getNom());
            return jeu;
        } catch (IOException | ClassNotFoundException e) {
            e.printStackTrace();
            System.out.println("Erreur lors du chargement.");
            return null;
        }
    }

    public static boolean hasSaves() {
        // Définir le répertoire des sauvegardes
        File savesDir = new File(SAVE_DIRECTORY_PATH);

        // Vérifier si le répertoire existe
        if (!savesDir.exists() || !savesDir.isDirectory()) {
            return false; // Aucun répertoire, donc aucune sauvegarde
        }

        // Lister les fichiers dans le répertoire
        File[] saveFiles = savesDir.listFiles((dir, name) -> name.endsWith(".jeu"));

        // Vérifier s'il y a des fichiers de sauvegarde
        return saveFiles != null && saveFiles.length > 0;
    }
}

File: KF-Remake/Kingdom-Fall/source/application/Sauvegarde.cs
using System;
using System.IO;
using Avalonia.Controls; // Pour OpenFileDialog et SaveFileDialog
namespace App;

public class Sauvegarde
{
    private static readonly string SAVE_DIRECTORY_PATH = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), "games/Kingdom-Fall/sauvegardes");

    public static async Task SauvegarderJeu(Jeu jeu, string nomJoueur)
    {
        var window = controllers.FenetreAppController.GetSingleton();
        var options = new Avalonia.Platform.Storage.FilePickerSaveOptions
        {
            Title = "Sauvegarder jeu",
            SuggestedFileName = nomJoueur + ".jeu",
            FileTypeChoices = new[]
            {
                new Avalonia.Platform.Storage.FilePickerFileType("Jeux") { Patterns = new[] { "*.jeu" } }
            }
        };

        DirectoryInfo directory = new DirectoryInfo(SAVE_DIRECTORY_PATH);
        Console.WriteLine(directory.FullName);
        if (!directory.Exists)
        {
            // Utilisation de StorageProvider pour sauvegarder le fichier
            var result = await window.StorageProvider.SaveFilePickerAsync(options);

            if (result != null)
            {
                try
                {
                    using (var fs = await result.OpenWriteAsync())
                    {
                        var json = System.Text.Json.JsonSerializer.Serialize(jeu);
                        using (var writer = new StreamWriter(fs))
                        {
                            writer.Write(json);
                        }
                        Console.WriteLine("Jeu sauvegardé avec succès.");
                    }
                }
                catch (IOException ex)
                {
                    Console.WriteLine("Erreur lors de la sauvegarde.");
                    Console.WriteLine(ex.ToString());
                }
            }
        }
    }

    public static async Task<Jeu> ChargerJeu()
    {
        var fileChooser = new OpenFileDialog
        {
            Title = "Charger jeu",
            Filters = { new FileDialogFilter { Name = "Jeux", Extensions = { "jeu" } } },
            Directory = SAVE_DIRECTORY_PATH,
            AllowMultiple = false
        };

        DirectoryInfo directory = new DirectoryInfo(SAVE_DIRECTORY_PATH);
        Console.WriteLine(directory.FullName);
        if (!directory.Exists)
        {
            directory.Create();
        }

        // Note : Vous devez passer une référence à la fenêtre principale (Avalonia Window)
        var files = await fileChooser.ShowAsync(controllers.FenetreAppController.GetSingleton());

        if (files != null && files.Length > 0)
        {
            try
            {
                using (FileStream fs = new FileStream(files[0], FileMode.Open))
                using (var reader = new StreamReader(fs))
                {
                    var json = reader.ReadToEnd();
                    Jeu jeu = System.Text.Json.JsonSerializer.Deserialize<Jeu>(json);
                    controllers.OptionsController.RenommerJoueur(jeu.joueur.nom);
                    return jeu;
                }
            }
            catch (Exception ex) when (ex is IOException || ex is System.Runtime.Serialization.SerializationException)
            {
                Console.WriteLine("Erreur lors du chargement.");
                Console.WriteLine(ex.ToString());
                return null;
            }
        }

        return null;
    }

    public static bool HasSaves()
    {
        DirectoryInfo savesDir = new DirectoryInfo(SAVE_DIRECTORY_PATH);

        if (!savesDir.Exists)
        {
            return false;
        }

        FileInfo[] saveFiles = savesDir.GetFiles("*.jeu");
        return saveFiles.Length > 0;
    }
}